dcn-oneflow/basemodel.py

```diff
@@ -0,0 +1,469 @@
+        :param shuffle: Boolean. Whether to shuffle the order of the batches at the beginning of each epoch.
+        :param callbacks: List of `deepctr_torch.callbacks.Callback` instances. List of callbacks to apply during training and validation (if ). See [callbacks](https://tensorflow.google.cn/api_docs/python/tf/keras/callbacks). Now available: `EarlyStopping` , `ModelCheckpoint`
+        :return: A `History` object. Its `History.history` attribute is a record of training loss values and metrics values at successive epochs, as well as validation loss values and validation metrics values (if applicable).
+        """
+        if isinstance(x, dict):
+            x = [x[feature] for feature in self.feature_index]
+
+        do_validation = False
+        if validation_data:
+            do_validation = True
+            if len(validation_data) == 2:
+                val_x, val_y = validation_data
+                val_sample_weight = None
+            elif len(validation_data) == 3:
+                val_x, val_y, val_sample_weight = validation_data  # pylint: disable=unpacking-non-sequence
+            else:
+                raise ValueError(
+                    'When passing a `validation_data` argument, '
+                    'it must contain either 2 items (x_val, y_val), '
+                    'or 3 items (x_val, y_val, val_sample_weights), '
+                    'or alternatively it could be a dataset or a '
+                    'dataset or a dataset iterator. '
+                    'However we received `validation_data=%s`' % validation_data)
+            if isinstance(val_x, dict):
+                val_x = [val_x[feature] for feature in self.feature_index]
+
+        elif validation_split and 0. < validation_split < 1.:
+            do_validation = True
+            if hasattr(x[0], 'shape'):
+                split_at = int(x[0].shape[0] * (1. - validation_split))
+            else:
+                split_at = int(len(x[0]) * (1. - validation_split))
+            x, val_x = (slice_arrays(x, 0, split_at),
+                        slice_arrays(x, split_at))
+            y, val_y = (slice_arrays(y, 0, split_at),
+                        slice_arrays(y, split_at))
+
+        else:
+            val_x = []
+            val_y = []
+        for i in range(len(x)):
+            if len(x[i].shape) == 1:
+                x[i] = np.expand_dims(x[i], axis=1)
+
+        train_tensor_data = Data.TensorDataset(
+            flow.from_numpy(
+                np.concatenate(x, axis=-1)),
+            flow.from_numpy(y))
+        if batch_size is None:
+            batch_size = 256
+
+        model = self.train()
+        loss_func = self.loss_func
+        optim = self.optim
+
+        if self.gpus:
+            print('parallel running on these gpus:', self.gpus)
+
+            # model = torch.nn.DataParallel(model, device_ids=self.gpus) 
```

Comment:
```suggestion
            # model = torch.nn.DataParallel(model, device_ids=self.gpus) 
```
这里oneflow里面没有找到相应的API，duop计算的问题

Reply:
这个是旧版的 pytorch 代码吧
https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel

Reply:
原来pytorch代码使用的是这个，我在[pytorch官网](https://pytorch.org/docs/stable/generated/torch.nn.DataParallel.html?highlight=torch%20nn%20dataparallel#torch.nn.DataParallel)看建议使用旧版的DistributedDataParallel。

Reply:
oneflow.nn.parallel.DistributedDataParallel这个我看文档里好像没有 `device_ids` 这个参数